lib/AST/ASTPrinter.cpp

```diff
@@ -523,7 +573,39 @@ class PrintAST : public ASTVisitor<PrintAST> {
       return;
     }
 
-    TL.getType().print(Printer, Options);
+    printCallback(TL.getType(), Printer, Options);
+  }
+
+  static void printWithParensIfNotSimple(Type ty, ASTPrinter &printer, PrintOptions &opts) {
+    if (!isSimple(ty)) {
+      printer << "(";
+      ty.print(printer, opts);
+      printer << ")";
+    } else {
+      ty.print(printer, opts);
+    }
+  }
+
+  void printTypeLocForImplicitlyUnwrappedOptional(TypeLoc TL) {
+    // Unwrap the underlying type and print that, followed by a
+    // '!'.
+
+    if (auto *refTy = TL.getType()->getAs<ReferenceStorageType>()) {
+      if (Options.PrintStorageRepresentationAttrs)
+        Printer << getStringForReferenceStorageTypeIntroducer(refTy);
+
+      auto objTy = refTy->getReferentType()->getAnyOptionalObjectType();
+      TL.setType(objTy);
+    } else if (auto inoutTy = TL.getType()->getAs<InOutType>()) {
+      assert(inoutTy->getObjectType()->getAnyOptionalObjectType());
+      TL.setType(InOutType::get(inoutTy->getObjectType()->getAnyOptionalObjectType()));
+    } else {
+      assert(TL.getType()->getAnyOptionalObjectType());
+      TL.setType(TL.getType()->getAnyOptionalObjectType());
+    }
+
+    printTypeLoc(TL, printWithParensIfNotSimple);
```

Comment:
Since this is the only place that uses this callback, I'd suggest just printing the parens here. Then `printTypeLoc` can go back to the way it was.

Also, hang on, the whole point of a TypeLoc is that it has a TypeRepr. If we're not printing the TypeRepr, can't we just jump straight to `printTransformedType`?

(Sorry to keep poking at this one!)